Add static type checking via RBS + Steep

lloeki opened the pull request:
<!--
Check out the
https://github.com/DataDog/dd-trace-rb/blob/master/docs/DevelopmentGuide.md
for guidance on how to set up your development environment,
run the test suite, write new integrations, and more.
-->

**What does this PR do?**

Add static type checking via RBS + Steep.

**Motivation**

Adding typing to dd-trace-rb has been discussed within the Ruby Guild, and initially attempted with Sorbet.

Unfortunately due to its focus coming from its Stripe ownership, Sorbet's scope and design is to typecheck applications, which doesn't lend itself well to libraries, and has technical limitations such as requiring possibly insecure side-effectful source evaluation or not being compatible with aarch64.

Sorbet's value is also limited without manual type annotations.

Static typing libddwaf-rb with RBS and Steep has found a dozen of edge-case issues, two of which would have had dramatic impact, proving the usage is promising at a larger scale.

This is therefore attempted with RBS syntax and Steep type checker, both projects being backed by the Ruby core team.

**Additional Notes**

This was prototyped during initial AppSec development, then removed by 8d05385654daa6b572fa180bd9fb538f9382c614. Lots has changed since then though.

**How to test the change?**

```
bundle exec steep check
```

ivoanjo commented:
This looks nice, thanks for taking a stab at it!

I see our spec related to "things we're shipping" needs to be updated -- is your plan to ship the sigs to customers with the first PR, or do that later?

If we're shipping this to customers, we should probably integrate steep into the usual CI test cycle, to make sure we don't regress.

If, on the other hand, we plan on working on this a bit before that, are there any downsides in merging this PR almost as-is?

marcotc commented:
I'm more than happy to support merging this as well, and helping with any tasks that ease our adoption of this static typing.

Echoing what @ivoanjo said, if we want to break it down, merging the foundation first then slowly working towards a fully fledged typing suite (for both ourselves and users), I'm more than happy to support incremental work even if there's no immediate value in each individual PR until the very end.

lloeki commented:
Fixed some more typing. My first goal will be to have `lib/datadog/appsec` fully typed. Still some work to do though, some of which will require refactoring.

Also, added a simple CI thing.

lloeki commented:
Ha, and Sorbet gets hopelessly confused by some of these syntax things xD

```
sorbet/rbi/hidden-definitions/hidden.rbi:3414: Redefining constant Datadog::AppSec::Configuration::DSL::Instrument https://srb.help/4012
    3414 |class Datadog::AppSec::Configuration::DSL::Instrument
    3415 |  def self.[](*arg); end
    3416 |
    3417 |  def self.members(); end
    3418 |end
    lib/datadog/appsec/configuration.rb:18: Previous definition
    18 |        Instrument = _ = Struct.new(:name, :options)
                ^^^^^^^^^^

sorbet/rbi/hidden-definitions/hidden.rbi:3420: Redefining constant Datadog::AppSec::Configuration::Settings::Integration https://srb.help/4012
    3420 |class Datadog::AppSec::Configuration::Settings::Integration
    3421 |  def self.[](*arg); end
    3422 |
    3423 |  def self.members(); end
    3424 |end
    lib/datadog/appsec/configuration/settings.rb:117: Previous definition
     117 |        Integration = _ = Struct.new(:integration, :options)
                  ^^^^^^^^^^^

lib/datadog/appsec/configuration.rb:28: Wrong number of arguments for constructor. Expected: 0, got: 2 https://srb.help/7004
    28 |          @instruments << Instrument.new(name, options)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Errors: 3
```

marcotc commented on lib/datadog/appsec/configuration.rb:
```diff
@@ -14,7 +14,9 @@ def self.included(base)
       # Configuration DSL implementation
       class DSL
-        Instrument = Struct.new(:name, :options)
+        # Struct constant whisker cast for Steep
```
What's is this `whisker` pattern here for Steep?

lloeki replied:
Assigning to `_` swallows typing information so that `_ = foo` always works. As a consequence `bar = _ = foo` acts as a forceful cast, ignoring the type of `foo` when assigning to `bar`.

This pattern should not be used lightly but is sometimes necessary. In this case it's because `Struct.new` returns a `Struct[untyped]` but here we want the `Instrument` constant to be of type `Instrument`. [See here](https://github.com/ruby/rbs/issues/627#issuecomment-1335085271) for details.

(The name comes from the looks of this assignment: `= _ =` looks like cat whiskers)

ivoanjo replied:
To be honest, this file is so big that I would just suggest just making it a regular `class` and not need any particular weird typing things :)

lloeki replied:
Yeah, I thought the same. I'll try to improve that on future quality iterations.

Overall I'll be trying to refactor all these things requiring syntax or comment directive oddities.

marcotc commented:
@lloeki, so we are aware how to work with this new tool, could you add a bit of documentation under https://github.com/DataDog/dd-trace-rb/tree/master/docs for how to use RBS/Steep in `ddtrace`?
Basic things like "how to type my new .rb file" would be nice, so we know how to expand type coverage.

lloeki commented:
Sure @marcotc! Let me fix the remaining non-typing issues so that it's not technically blocked and I will add something.

marcotc commented on lib/datadog/appsec/configuration/settings.rb:
```diff
@@ -126,37 +126,45 @@ def initialize
 
         def enabled
-          @options[:enabled]
+          # Cast for Steep
+          _ = @options[:enabled]
```
@lloeki is there no cleaner option for `_ = ...`? I ask because this assignment is not actually needed, and quite distracting.
I dream of a `@options[:enabled]` # steep:cast` or similar comment.

lloeki replied:
Sadly no: types appear to be assigned to each language construct + each value in a way that "pass through" many things, except `_`. So `@options` (and `@options[:enable]`) have consistent types throughout the codebase: there's no form of locality to typing.

Therefore the only way to cast is to go through an assignment to have a second type in there, which can normally go on the left hand side only to a type encompassing the one on the right hand side. To move from a non-encompassing type it needs to go through the special-typed `_` on the left hand side of the assignment, which "swallows" the type information. The new type is then "forced back" by either assignment or return type.

Even the following annotation does not work:

```
        def enabled
          # @type var _cast: untyped
          _cast = @options[:enabled]
        end
```

Indeed the `::Datadog::AppSec::Configuration::DSL::option` type still carries through:

```
lib/datadog/appsec/configuration/settings.rb:128:12: [error] Cannot allow method body have type `::Datadog::AppSec::Configuration::DSL::option` because declared as type `bool`
│   ::Datadog::AppSec::Configuration::DSL::option <: bool
│     ::Datadog::AppSec::Configuration::DSL::option <: (true | false)
│       (::Integer | bool | ::Symbol | ::String | ::StringIO | ::File | ::Regexp | ::Hash[(::Symbol | ::String), ::Datadog::AppSec::Configuration::DSL::option] | ::Array[::Datadog::AppSec::Configuration::DSL::option]) <: (true | false)
│         ::Integer <: (true | false)
│           ::Integer <: true
│
│ Diagnostic ID: Ruby::MethodBodyTypeMismatch
│
└         def enabled
              ~~~~~~~
```

As seen above, `@options[:enabled]` is a value of the `@options` hash, which has a very wide type to be able to store all sorts of types for all possible option keys, but we want the return type of each option to be stricter than the general case.

The only way around that is to have a fully typed, non-generic store, e.g a struct or an actual class instead of a dumb hash with dumb values.

ivoanjo replied:
Oh wow, I'm kinda with Marco, this is kinda distracting.

I don't think it's a blocker, but here's two suggestions:
* Replace the options actual instance variables. We already need quite a lot of boilerplate to access them inside options, so perhaps making them instance variables would reduce the boilerplate and make steep happy as a side-effect
* Remove the `# Cast for steep` (or maybe just leave the first one). It shows up a bit too often, so it really adds a lot of noise. And if someone thinks this is not needed and removes is, they'll quickly get reminded once they push to CI so it's not like they'll go far without learning it anyway :)

lloeki commented:
@marcotc @ivoanjo I just added a guide! LMKWYT

I think we should be all good now.

lloeki commented:
> is your plan to ship the sigs to customers with the first PR, or do that later?

Later. Signatures will start being directly useful for customers when we type the public API as it will e.g provide autocompletion as well as argument and return value type checking.

There are two complementary ways to approach progressive typing:

- start with precise typing of leafs (types that have little to no type dependencies) and go up from there
- start with precise typing of public API endpoints and go down from there

marcotc reviewed (approved):
The `docs/StaticTypingGuide.md` is very good! Thank you!
Let's give this a shot!

ivoanjo commented on Gemfile:
```diff
@@ -84,12 +84,20 @@ if RUBY_PLATFORM != 'java'
+  end
+
+  # type checking with steep
+  if RUBY_VERSION >= '2.6.0' && RUBY_PLATFORM != 'java'
```
Just out of curiosity -- does steep not run on JRuby?

lloeki replied:
I'm honestly not sure. I don't think it matters either as it's static analysis, like Rubocop.

ivoanjo commented on docs/StaticTypingGuide.md:
```diff
@@ -0,0 +1,193 @@
+
+Outputs `.rbs` content on stdout. The `source files` arguments lists the files from which the skeleton will be statically built from parsing. No evaluation occurs, therefore this has limited typing analysis capability, typically resulting in a lot of `untyped`.
+
+Note: Comments are reproduced as is which is useful for a visual check but should be manually removed when moving the skeleton to a `.rbs` file, because of the duplication and risk of comments getting desynced.
```
:+1: Thanks for calling this out, it's pretty important!

ivoanjo commented on docs/StaticTypingGuide.md:
```diff
@@ -0,0 +1,193 @@
+
+One can then proceed to [adjusting the signatures](https://github.com/ruby/rbs/blob/e91be7275f4005b1aeac8eadc2faa2b4ad5fdfef/docs/syntax.md) ([by example](https://github.com/ruby/rbs/blob/e91be7275f4005b1aeac8eadc2faa2b4ad5fdfef/docs/rbs_by_example.md)), removing as much `untyped` as possible.
+
+### Type profiling
+
+To discover types, one can leverage [`typeprof`](https://github.com/ruby/typeprof). Contrary to `rbs prototype rb` which relies solely on static parsing, `typeprof` is a Ruby interpreter, except it doesn't *execute* Ruby code, merely evaluates it to track types. Entry point calls to explore the various codepaths are required.
+
+With this file:
```
Do you know if there is any tool that gets types based on actual runs, not just abstract runs?

I'm thinking it'd be cool to have something like that to help me type the profiler native stuff. Otherwise, I'll just do it manually -- it's a bit of work but I don't have particularly complex types going in/out of native anyway.

lloeki replied:
There's `rbs prototype runtime` but turns out I never actually used it so I'm not entirely sure.

It's important to note that these are helpful in understanding typing and bootstrapping types, but ultimately it has to be reviewed by a human.

ivoanjo commented on lib/datadog/appsec/configuration/settings.rb:
```diff
@@ -8,7 +8,8 @@ module Configuration
         class << self
           def boolean
-            lambda do |v|
+            # @type ^(::String) -> bool
```
I'm curious -- why add the types inline here and not on an rbs?

ivoanjo commented on lib/datadog/appsec/extensions.rb:
```diff
@@ -112,6 +114,14 @@ def trace_rate_limit
           @settings.trace_rate_limit
         end
 
+        def obfuscator_key_regex
+          @settings.obfuscator_key_regex
+        end
+
+        def obfuscator_value_regex
+          @settings.obfuscator_key_regex
+        end
```
Wait, is this change (and a few others on this file) a new feature? It didn't quite look like this existed before.

lloeki replied:
It was actually an overlook, which typing pointed me to! So it's a consistency fix for an existing feature.

ivoanjo commented on sig/datadog/appsec/contrib/auto_instrument.rbs:
```diff
@@ -0,0 +1,11 @@
+module Datadog
+  module AppSec
+    module Contrib
+      # Auto-instrumentation for security integrations
+      # TODO: this implementation is trivial, check for shareable code with tracer
```
Leftover comment ;)

Should we maybe have it part of the CI check that there are no comments here, or is that a bit too agressive?

ivoanjo replied:
(There's a few other .rbs files with comments as well)

lloeki replied:
I think we do want comments in `.rbs`, only related to typing, e.g to explain why this type or alias is that way, or TODOs, or... you get the point!

lloeki replied:
(Good catch though)

ivoanjo replied:
Good point. I guess we can keep an eye out for "oops this comment should not be in" during reviews and at some point it should get pretty rare because we'll not be creating new `.rbs` files very often.

ivoanjo commented on sig/datadog/appsec/contrib/rails/ext.rbs:
```diff
@@ -0,0 +1,14 @@
+    module Contrib
+      module Rails
+        # Rack integration constants
+        module Ext
+          APP: "rails"
+
+          ENV_ENABLED: "DD_TRACE_RAILS_ENABLED"
+        end
```
Does steep complain if we change these constants (especially given we're not checking this folder), or do we need to remember to keep them up-to-date?

lloeki replied:
I have to check but I think it does. This can also happen on method arguments and return values that have a known constant value or list of values possible.

When I wrote some of these by hand I put the type in instead of the value though. This is one of the things about which I'm not sure yet what would be best. e.g does Steep check for such static values at call sites and for method body returns, kind of like typed C/Go enums? I can see that being useful balanced against the value duplication.

ivoanjo commented on sig/datadog/appsec/waf.rbs:
```diff
@@ -0,0 +1,230 @@
+module Datadog
+  module AppSec
+    module WAF
+      module LibDDWAF
```
Should this perhaps go inside libddwaf directly?

lloeki replied:
It actually exists in libddwaf, but its `sig` isn't currently in the gem. I went the simple way first, and plan to address that ultimately.

ivoanjo commented on lib/datadog/appsec/utils/http/media_range.rb:
```diff
@@ -0,0 +1,201 @@
+    module Utils
+      module HTTP
+        # Implementation of media range for content negotiation
+        class MediaRange
+          class ParseError < ::StandardError
+          end
```
Wow this is a big addition just to support `Response#format`. I hesistate to ask but... is there any way we could we keep some of the simplicity of the old code?

lloeki replied:
Yeah, I tried. The thing is, the old code was not correct WRT what real-life UAs might send.

So ultimately my choice was either to pile up some more hacks that may still be incomplete in face of whatever UAs do, or implement the real thing.

This is also on a feature that is implemented but currently gated by an undocumented, experimental setting, so the actual risk to the change here is low.

ivoanjo reviewed (commented):
Github doesn't let me comment on empty files but there's a few:

* `sig/datadog/appsec/contrib/rails/integration.rbs`
* `sig/datadog/appsec/contrib/sinatra/integration.rbs`

-- is that expected?

Other than that, I gave it a quick pass and it looks reasonable.

Sorry for not having time to do a full pass, hopefully this still helps :)
On my side I see no reason not to merge this :+1:
